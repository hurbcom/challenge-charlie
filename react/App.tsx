import React, { Suspense } from "react";
import ErrorBoundary from "./ErrorBoundary";
import { GlobalStyle } from "./styles/global";
import HeroImage from "@components/HeroImage";
import WeatherAccordeon from "@components/WeatherAccordeon";

const App: React.FC = () => {
    return (
        <Suspense fallback={<div>Suspending...</div>}>
            <ErrorBoundary>
                <GlobalStyle />
                <HeroImage
                    image={
                        "https://www.bing.com/th?id=OHR.KitsAspen_PT-BR8299899730_1920x1080.jpg&rf=LaDigue_1920x1080.jpg&pid=hp"
                    }
                >
                    <WeatherAccordeon />
                </HeroImage>
            </ErrorBoundary>
        </Suspense>
    );
};

export default App;
