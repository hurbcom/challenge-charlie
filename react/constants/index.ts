export const POSITIONS_TO_OPACITIES: Record<number, string> = {
    0: "--max-bg-opacity",
    1: "--medium-bg-opacity",
    2: "--min-bg-opacity",
};

export const SIZE = {
    mobileS: "320px",
    mobileM: "375px",
    mobileL: "425px",
    tablet: "768px",
    laptop: "1024px",
    laptopL: "1440px",
    desktop: "2560px",
};

export const DEVICES = {
    mobileS: `(min-width: ${SIZE.mobileS})`,
    mobileM: `(min-width: ${SIZE.mobileM})`,
    mobileL: `(min-width: ${SIZE.mobileL})`,
    tablet: `(min-width: ${SIZE.tablet})`,
    laptop: `(min-width: ${SIZE.laptop})`,
    laptopL: `(min-width: ${SIZE.laptopL})`,
    desktop: `(min-width: ${SIZE.desktop})`,
    desktopL: `(min-width: ${SIZE.desktop})`,
};
