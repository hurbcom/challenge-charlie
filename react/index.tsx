import React, { StrictMode } from "react";
import { hydrateRoot } from "react-dom/client";
import App from "./App";

const root = document.getElementById("react-app");

if (root) {
    hydrateRoot(
        root,
        <StrictMode>
            <App />
        </StrictMode>
    );
}
