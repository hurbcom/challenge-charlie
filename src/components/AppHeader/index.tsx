import { FormEvent, useEffect, useRef, useState } from "react";
import styled from "styled-components";
import Icon from "components/Icon";
import useGeolocation from "hooks/useGeolocation";
import useGeocode, { AppCoordinatesType } from "hooks/useGeocode";

const AppHeaderBase = styled.div`
    background-color: rgba(255, 255, 255, 0.9);
    color: #212121;
    display: flex;
    padding: 8px;
    gap: 8px;
    align-items: center;
    color: #8d8987;

    &:focus-within {
        background-color: rgba(255, 255, 255, 1);
    }
`;

const Input = styled.input`
    border: none;
    padding: 8px;
    display: flex;
    flex-grow: 1;
    border: none;
    color: inherit;
    background-color: transparent;
`;

export type AppHeaderType = {
    onChange: (coords: AppCoordinatesType) => void;
};

const AppHeader = ({ onChange }: AppHeaderType) => {
    const { translate, result } = useGeocode();
    const searchInputRef = useRef<HTMLInputElement>(null);
    const [query, setQuery] = useState<string>("");
    const { coords } = useGeolocation();
    const handleOnChange = (event: FormEvent<HTMLInputElement>) => {
        setQuery(event.currentTarget.value);
    };

    useEffect(() => {
        if (!result?.latitude || !result?.longitude) {
            return;
        }
        onChange({
            latitude: result.latitude,
            longitude: result.longitude,
        });
    }, [result]);

    useEffect(() => {
        if (!coords?.latitude || !coords?.longitude) {
            return;
        }
        translate({
            search: {
                latitude: coords.latitude,
                longitude: coords.longitude,
            },
        });
    }, [coords]);

    useEffect(() => {
        if (query.length > 2) {
            translate({ search: query });
        }
    }, [query]);

    return (
        <AppHeaderBase role="search">
            <Icon name="compass" size="large" />
            <Input
                type="text"
                ref={searchInputRef}
                placeholder="Buscar um lugar..."
                value={query}
                onChange={handleOnChange}
            />
        </AppHeaderBase>
    );
};

export default AppHeader;
