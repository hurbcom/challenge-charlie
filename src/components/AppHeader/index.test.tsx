import { act, fireEvent, render, screen } from "@testing-library/react";
import AppHeader from ".";

const handleChange = jest.fn();

jest.mock("../../hooks/useDailyForecast", () => {
    return jest.fn(() => ({
        result: {
            latitude: -27.1,
        },
        loadForecast: jest.fn().mockImplementation(() => Promise.resolve()),
    }));
});

jest.mock("../../hooks/useGeocode", () => {
    return jest
        .fn()
        .mockImplementation(() => ({
            result: {},
            translate: jest.fn(),
        }))
        .mockImplementationOnce(() => ({
            result: {
                latitude: -27.1,
            },
            translate: jest.fn().mockImplementation(() =>
                Promise.resolve({
                    latitude: 27.5,
                    longitude: 48.5,
                    city: "Florianópolis",
                    state: "Santa Catarina",
                })
            ),
        }))
        .mockImplementationOnce(() => ({
            result: {
                latitude: -27.1,
                longitude: -48.5,
            },
            translate: jest.fn().mockImplementation(() =>
                Promise.resolve({
                    latitude: -27.5,
                    longitude: -48.5,
                    city: "Florianópolis",
                    state: "Santa Catarina",
                })
            ),
        }));
});

jest.mock("../../hooks/useGeolocation", () => {
    return jest.fn().mockImplementation(() => ({
        coords: {
            latitude: 27.5,
            longitude: 48.5,
        },
    }));
});

describe("<AppHeader />", () => {
    test("renders the <AppHeader/> component with users location", async () => {
        const { asFragment, rerender } = render(
            <AppHeader onChange={handleChange} />
        );
        expect(asFragment).toMatchSnapshot();
    });

    test("renders the <AppHeader/> component", async () => {
        const { asFragment } = render(<AppHeader onChange={handleChange} />);
        expect(asFragment).toMatchSnapshot();
        const input =
            screen.getByPlaceholderText<HTMLInputElement>("Buscar um lugar...");
        expect(input).toBeInTheDocument();

        await act(async () => {
            fireEvent.change(input, { target: { value: "floripa" } });
        });

        expect(input.value).toEqual("floripa");
    });
});
