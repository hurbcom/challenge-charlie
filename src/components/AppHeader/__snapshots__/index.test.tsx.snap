// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppHeader /> renders the <AppHeader/> component 1`] = `[Function]`;

exports[`<AppHeader /> renders the <AppHeader/> component with users location 1`] = `[Function]`;
